In [75]:
import numpy as np
import pandas as pd

# Question 1

What's the version of NumPy that you installed?

In [76]:
np.__version__

'1.23.1'

### Answer 1 :

The version of numpy that I installed in my pc is 1.23.1

In [77]:
data = pd.read_csv('carprice.csv')

# Question 2

How many records are in the dataset?

Here you need to specify the number of rows.

In [78]:
data.head(5)

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [79]:
data.shape

(11914, 16)

## Answer 2

The number of row in this dataset is 11914 rows

# Question 3

Who are the most frequent car manufacturers (top-3) according to the dataset?

In [80]:
data['Make'].value_counts().head(3)

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

# Answer 3

Top 3 car manufactured is Chevrolet, Ford and Volkwagen

# Question 4

What's the number of unique Audi car models in the dataset?

In [81]:
data_1 = data[data['Make'] == 'Audi']

In [82]:
data_1['Model'].nunique()

34

## Answer 4

Numer of unique Audi car model in the dataset is 34

# Question 5

How many columns in the dataset have missing values?

In [83]:
data_2 = data.isna().sum()
data_2

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

## Answer 5

There is 5 missing values in this dataset

# Question 6

Does median value change after filling missing value



In [84]:
# looking for median value
data['Engine Cylinders'].median()

6.0

In [85]:
#looking for the mode
data_mode = data['Engine Cylinders'].mode()[0]
data_mode

4.0

In [90]:
 #fill the na in Engine Cylinders columns with mode
data['Engine Cylinders'].fillna(data_mode, inplace= True)

In [91]:
data['Engine Cylinders'].isna().sum()

0

In [92]:
# looking for median value
data['Engine Cylinders'].median()

6.0

## Answer 6

Median does not change afer fill with mode, question is `no`

# Question 7

Questions 7: Value of the first element of w


In [93]:
data_3 = data[data['Make'] == 'Lotus']

In [95]:
data_3.head(3)

Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact   Convertible           27   
3913  Exotic,High-Performance      Compact   Convertible           26   
3914  Exotic,High-Performance      Compact   Convertible           27   

      city mpg  Popularity   MSRP  
3912        21         613  43995  
3913        20         613  54990  
3914        21         613  47250

In [102]:
new_data = data_3[['Engine HP', 'Engine Cylinders']]
new_data.shape

(29, 2)

In [106]:
new_data.drop_duplicates(inplace= True)

C:\Users\Titim\AppData\Local\Temp\ipykernel_10100\843009859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data.drop_duplicates(inplace= True)


In [107]:
new_data.shape

(9, 2)

In [108]:
new_data

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [132]:
X = new_data.to_numpy()

In [133]:
XTX = np.dot(X.T, X)

In [134]:
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [135]:
inv = np.linalg.inv(XTX)

In [136]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [143]:
  z = np.dot(inv, X.T)

In [144]:
w = np.dot(z, y)

In [145]:
w

array([  4.59494481, -63.56432501])

## Answer 7

The first element of w is  4.59494481